src/hpr_packet_reporter.erl

```diff
@@ -0,0 +1,302 @@
+        | integer()
+        | undefined
+}).
+
+%%%===================================================================
+%%% API Functions
+%%%===================================================================
+
+start_link(Args) ->
+    gen_server:start_link({local, ?SERVER}, ?SERVER, Args, []).
+
+-spec init_ets() -> ok.
+init_ets() ->
+    ?ETS = ets:new(?ETS, [named_table, bag, public, {write_concurrency, true}]),
+    ok.
+
+-spec report_packet(hpr_packet_up:packet() | #packet_router_packet_down_v1_pb{}) -> ok.
+report_packet(Packet) ->
+    EncodedPacket = encode_packet(Packet),
+    Timestamp = erlang:system_time(millisecond),
+    true = ets:insert(?ETS, {Timestamp, EncodedPacket}),
+    ok.
+
+-spec write_packets() -> ok.
+write_packets() ->
+    gen_server:cast(?SERVER, write).
+
+-spec upload_packets(string()) -> ok.
+upload_packets(FilePath) ->
+    gen_server:cast(?SERVER, {upload, FilePath}).
+
+%%%===================================================================
+%%% Spawning and gen_server implementation
+%%%===================================================================
+
+init(#{base_dir := BaseDir} = _Args) ->
+    AWSClient = setup_aws(),
+    FilePath = filename:join(BaseDir, "tmp/"),
+    ok = filelib:ensure_dir(FilePath),
+    start_report_interval(),
+    {ok, #state{
+        aws_client = AWSClient,
+        base_dir = FilePath
+    }}.
+
+handle_call(_Request, _From, State = #state{}) ->
+    {reply, ok, State}.
+
+handle_cast(write, State = #state{base_dir = BaseDir}) ->
+    Timestamp = erlang:system_time(millisecond),
+    %% TODO: File rotation
+    FilePath = filename:join(BaseDir, "packetreport.gz"),
+
+    {ok, S} = open_tmp_file(FilePath),
+    Data = get_packets_by_timestamp(Timestamp),
+
+    lists:foreach(
+        fun(Packet) ->
+            io:fwrite(S, "~w~n", [Packet])
+        end,
+        Data
+    ),
+    file:close(S),
+
+    NumDeleted = delete_packets_by_timestamp(Timestamp),
+    FileSize = filelib:file_size(FilePath),
+    MaxFileSize = application:get_env(hpr, packet_reporter_max_file_size, ?MAX_FILE_SIZE),
+    lager:info(
+        [
+            {packets_processed, length(Data)},
+            {packets_deleted, NumDeleted},
+            {file_size, FileSize},
+            {max_file_size, MaxFileSize}
+        ],
+        "packet reporter processing"
+    ),
+
+    case FileSize >= MaxFileSize of
+        true -> ?MODULE:upload_packets(FilePath);
+        false -> skip
+    end,
+    {noreply, State};
+handle_cast({upload, FilePath}, State = #state{aws_client = AWSClient}) ->
+    Timestamp = erlang:system_time(millisecond),
+    FileName = "packetreport." ++ integer_to_list(Timestamp) ++ ".gz",
+    upload_file(AWSClient, FilePath, FileName),
+    file:delete(FilePath),
+    {noreply, State};
+handle_cast(_Request, State = #state{}) ->
+    {noreply, State}.
+
+handle_info(_Info, State = #state{}) ->
+    {noreply, State}.
+
+terminate(_Reason, _State = #state{}) ->
+    ok.
```

Comment:
We may want to look at the reason here and try to write the last of what's in ETS to file before we shut down and lose it.